### DISCUSSION TITLE
How to setup CustomBinding from legacy config

### DISCUSSION THREAD

**Discussion by OlegSharov1** [CoreWCF/CoreWCF#1327] [Q&A]
Hello,

I'm currently transitioning our server-side WCF to CoreWCF, but I'm encountering an issue with setting up the host, similar to how it was configured in ServiceModel. The original host configuration is as follows:

```
<binding name="myCustomBinding" sendTimeout="01:00:00" closeTimeout="00:01:00" openTimeout="00:01:00" receiveTimeout="00:01:00">
    <security authenticationMode="UserNameOverTransport" requireSecurityContextCancellation="false">
        <secureConversationBootstrap authenticationMode="AnonymousForSslNegotiated"></secureConversationBootstrap>
    </security>
    <sslStreamSecurity requireClientCertificate="false" />
    <tcpTransport maxReceivedMessageSize="2000000000" connectionBufferSize="250000" maxBufferSize="2000000000" />
</binding>
```

This myCustomBinding is then used here:

```
      <endpoint name="MyEndpoint" binding="customBinding" bindingConfiguration="myCustomBinding" contract="IMyContract" address="net.tcp://127.0.0.1">
      </endpoint>
```

Could you provide any guidance or ideas on how to programmatically replicate this configuration of MyEndpoint in CoreWCF? Thanks a lot!

**Comment by mconnew** [comment]
I initially thought this would be supported as on its face it looks like this is `TransportWithMessageCredentials`, but when I tried to write the code and convert each piece from config to code, I discovered that the secureConversationBootstrap config you are using requires full `Message` security. The authentication mode of `AnonymousForSslNegotiated` is an odd one, it basically uses SOAP messages as the carrier transport for completing a TLS handshake between the client and the server. Once a TLS session is established, a session encryption key is communicated across the secure TLS session, so now each end knows a symmetric encryption key created just for this session, and the TLS session is discarded. Then it uses the symmetric encryption key using full Message security to send a request for a WS-SecureConversation token to be issued, authenticating with the username and password. Because you have set requireSecurityContextCancellation to false, an encrypted cookie is returned to the client which contains all the claims and authentication info the service needs. This way when the service receives the cookie, it can decrypt it and know all about the client authenticated identity without having to hold onto any state information locally. Once it has a WS-SecureConversation token (containing this cookie), the session key and the channel used is all discarded and it makes a new connection to the service which is secured using a TLS connection at the transport level. This communication is now using `TransportWithMessageCredentials` The server provides a certificate as is normal with TLS, but the client does not provide a certificate. Over this secure connection, the client makes calls to operations and the messages are authenticated using the WS-SecureConversation token obtained earlier. On receiving a request, the service decrypts the cookie from the token to obtain all the claims info and sets that as the request identity, then dispatches the request.  

What's odd/interesting is that you are doing this with TCP. The use of cookies (not requiring cancellation of session) is normally used with HTTP as you don't have a persistent connection and each request could land on a different server behind a load balancer. This doesn't apply to NetTcp which does have a persistent connection, so this is unnecessary. When the client socket disconnects, you can't reconnect and reuse the same WS-SecureConversation token with the service, you have to start again.  

When using NetTcp, you normally own both ends of the communication. Is this the case for you? Do you have scope to use a different binding which is more efficient, provides the same level of security, and is supported in CoreWCF? Migration can be relatively smooth. The steps are:
1. Add additional endpoint to existing WCF service using new binding
2. Modify the client app to use the new binding and endpoint
3. Once all clients are using the new endpoint, remove the old one from the service
4. Now you can replace the service with CoreWCF as you have a compatible endpoint

If you are able to use a different binding, let me know and I'll provide you with my recommendation.

**Reply by OlegSharov1** [reply]
First off, I want to thank you for the comprehensive response, I really appreciate it.

I want to apology because I made a mistake in my question, I provided wrong configuration.
In our actual configuration, parameter requireSecurityContextCancellation is set to true and not false, as I mentioned initially.

We also use a custom binding element, where we override some behaviour e.g. we use x-deflate header.
But it seems there is no problem with this element, it will be shown later.

I am going to describe the code I have created so far:

1. Create NetTcpBinding and set some parameters.
```
  var netTcp = new NetTcpBinding(SecurityMode.TransportWithMessageCredential);
  netTcp.Security.Transport.ClientCredentialType = TcpClientCredentialType.None;
  netTcp.Security.Message.ClientCredentialType = MessageCredentialType.UserName;
  netTcp.SendTimeout = TimeSpan.FromHours(1);
  netTcp.CloseTimeout = TimeSpan.FromMinutes(1);
  netTcp.OpenTimeout = TimeSpan.FromMinutes(1);
  netTcp.ReceiveTimeout = TimeSpan.FromMinutes(1);

  netTcp.MaxBufferSize = int.MaxValue;
  netTcp.MaxReceivedMessageSize = long.MaxValue;
  netTcp.MaxBufferPoolSize = int.MaxValue;
  netTcp.MaxConnections = int.MaxValue;
  netTcp.ReaderQuotas = new System.Xml.XmlDictionaryReaderQuotas()
  {
      MaxDepth = int.MaxValue,
      MaxArrayLength = int.MaxValue,
      MaxBytesPerRead = int.MaxValue,
      MaxNameTableCharCount = int.MaxValue,
      MaxStringContentLength = int.MaxValue,
  };
```

2. Create binding elements from this NetTcpBinding which can be added to the custom binding.
`var elements = netTcp.CreateBindingElements();`

3. Create custom binding and add elements into it. As you can see, I also add CoreWcfCompressionMessageEncodingBindingElement which inherits from MessageEncodingBindingElement and is responsible for message compression.

```
   var binding = new CustomBinding();
   binding.Elements.Add(new CoreWcfCompressionMessageEncodingBindingElement(new BinaryMessageEncodingBindingElement(), CompressionAlgorithm.Deflate));
  // Adding all the binding elements from NetTcp
   foreach (var element in elements)
   {
       binding.Elements.Add(element);
   }
```

So with this configuration, I can see in the overriden ReadMessage method, that message was successfully sent from the client and was decompressed on the server side without any problem. But just after, in the overriden SendMessage method I can see WCF is sending message to the client (after I added IncludeExceptionDetailInFaults to the serviceOptions):

The message could not be processed. This is most likely because the action 'http://tempuri.org/IMyService/Mymethod' is incorrect or because the message contains an invalid or expired security context token or because there is a mismatch between bindings. The security context token would be invalid if the service aborted the channel due to inactivity. To prevent the service from aborting idle sessions prematurely increase the Receive timeout on the service endpoint's binding.

At this point, I do not know what configuration should be modified to make it work.

I just know that if I comment out all the security-related attributes from the client config and set NetTcpBinding(SecurityMode.None) on the server side of CoreWCF, communication works.

Do you have any idea what could be set on the server side to make it work with security? Thank you

**Reply by OlegSharov1** [reply]
I am able to use different binding which is compatible with CoreWCF. Could you give me reccomendation please, which one would be suitable with the same security level?

**Reply by mconnew** [reply]
The binding you are using in your latest code snippet provides equivalent security to what you were using earlier. We changed the api surface of MessageEncoder slightly in CoreWCF and I want to rule out any issues due to any changes needed there that being the cause of the failure. Can you test with using the nettcp binding without adding CoreWcfCompressionMessageEncodingBindingElement and see if that works? I'm sure it's not needed to be said, but just to remove any uncertainty, you need to change to this binding at both ends because the old and new binding are equivalent from a security standpoint, but the on-wire communication is different. If it doesn't work, then I should be able to repro it locally. If it does, then there's some issue with CoreWcfCompressionMessageEncodingBindingElement. If you are able to share that, it would make it a lot easier to diagnose.

**Reply by OlegSharov1** [reply]
CoreWcfCompressionMessageEncodingBindingElement is an old implementation of Deflate/GZip compression format, when it was not possible to just set CompressionFormat.Deflate (before .NET Framework 4.5 I guess). Without this element, there is no problem with tcp bidning.

I am going to listen to your recommendation and will add additional new endpoints with new bindings, update client apps and then remove old endpoints.

I just have last question. 

Our original wcf hosts are configured and started in System.ServiceModel way. Lets say I configure and then open host like this:

`myServiceHost.Open()`

So I have multiple services, let's say: 

1. net.tcp://127.0.0.1:2026/FirstService
2. net.tcp://127.0.0.1:2026/SecondService

Idea is to add new endpoints with new bindings, so additionaly I would have:

3. net.tcp://127.0.0.1:2026/FirstServiceCoreWcf
4. net.tcp://127.0.0.1:2026/SecondServiceCoreWcf

which implement the same contracts using new bindings.

However, is it possbile to keep FirstService and SecondSevice as "ServiceModel" services and FirstServiceCoreWcf and SecondServiceCoreWcf as "CoreWCF"?

The question is about initialization, new CoreWCF services are started like this:

```
        public void StartServices()
        {
            Task.Run(() =>
            {
                CreateWebHostBuilder(new string[] { }).Build().Run();
            });
        }

        public static IWebHostBuilder CreateWebHostBuilder(string[] args) =>
            WebHost.CreateDefaultBuilder(args)
            .UseNetTcp(IPAddress.Loopback, 2026)
            .UseStartup<Startup>();
			
        public class Startup
        {
            public void ConfigureServices(IServiceCollection services)
            {
                services.AddServiceModelServices();
                services.AddServiceModelMetadata();
            }

            public void Configure(IApplicationBuilder app)
            {
                app.UseServiceModel(serviceBuilder =>
                {
		    serviceBuilder.AddService<FirstService>(serviceOptions =>
                    {
                        serviceOptions.DebugBehavior.IncludeExceptionDetailInFaults = true;
                    });

                    var mainServiceBinding = createCompressBinding();
                    var mainServiceName = "FirstServiceCoreWcf";

                    serviceBuilder.AddServiceEndpoint<FirstService, IFirstService>(mainServiceBinding, "/" + mainServiceName, endpoint =>
                    {
                        endpoint.Name = mainServiceName;

                    });
					
					// Add SecondService etc..
			    }
		    }
		}
```
This setting does not allow me to connect to original ServiceModel endpoints running on the same port 2026 (net.tcp://127.0.0.1:2026/FirstService and net.tcp://127.0.0.1:2026/SecondService) , I guess because of  .UseNetTcp(IPAddress.Loopback, 2026)? 

Is there any workaround how to make it work?

Thank you.
